hyperspace/cosmos/src/provider.rs

```diff
@@ -20,27 +22,95 @@ use ibc_proto::{
 use ibc_rpc::PacketInfo;
-use ics07_tendermint::{client_message::Header, events::try_from_tx};
-use pallet_ibc::light_clients::{AnyClientState, AnyConsensusState};
-use primitives::{Chain, IbcProvider, UpdateType};
+use ics07_tendermint::{client_message::Header, events::try_from_tx,
+					   client_message::ClientMessage, merkle::convert_tm_to_ics_merkle_proof
+};
+use pallet_ibc::light_clients::{AnyClientMessage, AnyClientState, AnyConsensusState};
+use primitives::{Chain, IbcProvider, UpdateType, mock::LocalClientTypes};
 use std::{pin::Pin, str::FromStr};
-use tendermint_rpc::{query::Query, Client, Order};
+use tendermint::block::{Height as BlockHeight, Header as BlockHeader};
+use tendermint_light_client::components::io::{AsyncIo, AtHeight};
+use tendermint_proto::Protobuf;
+use tendermint_rpc::{query::{EventType, Query}, event::{Event, EventData}, Client, Order,
+					 SubscriptionClient, WebSocketClient, abci::Path as TendermintABCIPath
+};
+use ibc::{
+	core::ics02_client::msgs::update_client::MsgUpdateAnyClient, tx_msg::Msg,
+	keys::STORE_KEY,
+};
+use crate::utils::{
+	event_is_type_channel, event_is_type_client,
+	event_is_type_connection, ibc_event_try_from_abci_event,
+};
+
+const KeyClientStorePrefix: &str = "clients";
 
 pub enum FinalityEvent {
-	Tendermint(Header),
+	Tendermint(BlockHeader),
 }
 
 pub struct TransactionId<Hash> {
 	pub hash: Hash,
 }
 
+impl<H> CosmosClient<H>
+where
+	H: Clone + Send + Sync + 'static
+{
+	async fn msg_update_client_header(&mut self, trusted_height: Height) -> Result<(Header, UpdateType), anyhow::Error> {
+		let latest_light_block = self.light_provider.fetch_light_block(AtHeight::Highest).await?;
+		let height = BlockHeight::try_from(trusted_height.revision_height)?;
+		let trusted_light_block = self.light_provider.fetch_light_block(AtHeight::At(height)).await?;
+
+		let update_type = match latest_light_block.validators == latest_light_block.next_validators {
+			true => UpdateType::Optional,
+			false => UpdateType::Mandatory
+		};
+
+		Ok((Header{
+			signed_header: latest_light_block.signed_header,
+			validator_set: latest_light_block.validators,
+			trusted_height,
+			trusted_validator_set: trusted_light_block.validators
+		}, update_type))
+	}
+
+	async fn query_tendermint_proof(&self, key: Vec<u8>, height: Height) -> Result<(Vec<u8>, Vec<u8>), Error>{
+		let path = format!("store/{}/key", STORE_KEY).as_str()
+			.parse::<TendermintABCIPath>().map_err(|err| Error::Custom(format!("failed to parse path: {}", err)));
+		let height = BlockHeight::try_from(height.revision_height);
+		let query_res =  self.rpc_client.abci_query(
+			path.ok(),
+			key,
+			height.ok(),
+			true,
+		).await?;
+
+		if !query_res.code.is_ok() {
+			// Fail with response log.
+			// todo()! add response code to error
+			return Err(Error::Custom(format!("failed abci query")));
+		}
+
+		if query_res.proof.is_none() {
+			// Fail due to empty proof
+			return Err(Error::Custom(format!("proof response is empty")));
```

Comment:
alternatively you can express this as a `match`

Reply:
wouldn't match result to more lines of code since I'd have to use different match statements for both checks? One checks for code and the other for proof.

Reply:
The end result is the same, it's just a matter of style. But I would have done it like that. 
Let's assume that query_res is a `QueryResult {proof: Option<T>, code: Result<V>}`
Then you could
```rs
match query_result {
      QueryResult {
           code: Ok(_), ...
      } => return Err(Error::Custom(format!("failed abci query"))),
      QueryResult {
           proof: None, ...
      } => return Err(Error::Custom(format!("failed abci query"))),
}

Reply:
wouldn't match result to more lines of code since I'd have to use different match statements for both checks? One checks for code and the other for proof.

Reply:
noted